migrations/20230104225718-transaction-balances-order-by-create-date.js

```diff
@@ -0,0 +1,189 @@
+          coalesce(disputed."netAmountInHostCurrency", 0) "disputedNetAmountInHostCurrency",
+          tb."hostCurrency"
+        FROM "TransactionBalances" tb
+        INNER JOIN "CollectiveBalanceCheckpoint" cbc ON tb."id" = cbc."id"
+        LEFT JOIN LATERAL (
+          SELECT
+            SUM(t."amountInHostCurrency") +
+              SUM(coalesce(t."platformFeeInHostCurrency", 0)) +
+              SUM(coalesce(t."hostFeeInHostCurrency", 0)) +
+              SUM(coalesce(t."paymentProcessorFeeInHostCurrency", 0)) +
+              SUM(coalesce(t."taxAmount" * t."hostCurrencyFxRate", 0)) "netAmountInHostCurrency"
+          FROM "Transactions" t
+          WHERE t."CollectiveId" = tb."CollectiveId"
+            AND t.id > tb."id"
+            AND t."deletedAt" is null
+          GROUP by t."CollectiveId"
+        ) as t ON TRUE
+        LEFT JOIN LATERAL (
+          SELECT
+            SUM(t."amountInHostCurrency") +
+              SUM(coalesce(t."platformFeeInHostCurrency", 0)) +
+              SUM(coalesce(t."hostFeeInHostCurrency", 0)) +
+              SUM(coalesce(t."paymentProcessorFeeInHostCurrency", 0)) +
+              SUM(coalesce(t."taxAmount" * t."hostCurrencyFxRate", 0)) "netAmountInHostCurrency"
+          FROM "Transactions" t
+          where t."CollectiveId" = tb."CollectiveId"
+            AND t."deletedAt" is null
+            AND t."isDisputed"
+            AND t."RefundTransactionId" is null
+          GROUP BY t."CollectiveId"
+        ) as disputed ON TRUE
+      );
+    `);
+}
+
+module.exports = {
+  async up(queryInterface) {
+    // Remove dependent objects of TransactionBalances materialized view
+    await removeDependentObjectsTransactionBalances(queryInterface);
+
+    /* START: Update TransactionBalances materialized view */
+    await queryInterface.sequelize.query(`DROP MATERIALIZED VIEW "TransactionBalances"`);
+
+    await queryInterface.sequelize.query(`
+      CREATE MATERIALIZED VIEW "TransactionBalances" AS (
+        WITH "ActiveCollectives" AS (
+          SELECT t."CollectiveId" as "ActiveCollectiveId"
+          FROM "Transactions" t
+          LEFT JOIN  "Collectives" c ON c."id" = t."CollectiveId" AND c."deletedAt" IS NULL
+          WHERE t."deletedAt" IS NULL AND t."hostCurrency" IS NOT NULL
+          AND c."isActive" IS TRUE
+          GROUP BY t."CollectiveId"
+          HAVING COUNT(DISTINCT t."hostCurrency") = 1
+        )
+        SELECT
+          "id",
+          "CollectiveId",
+          "createdAt",
+          "hostCurrency",
+          SUM(
+            COALESCE("amountInHostCurrency", 0)
+            + COALESCE("platformFeeInHostCurrency", 0)
+            + COALESCE("hostFeeInHostCurrency", 0)
+            + COALESCE("paymentProcessorFeeInHostCurrency", 0)
+            + COALESCE("taxAmount" * "hostCurrencyFxRate", 0)
+          ) OVER (PARTITION BY "CollectiveId", "hostCurrency" ORDER BY "createdAt") as "balance"
```

Comment:
Is this working? 

It seemed yesterday that `ORDER BY "createdAt"` was not enough and we needed `ORDER BY "createdAt", "id"`. I don't have the explanation at this point, would be interesting to have the input from @hdiniz 

Example:

```
        SELECT
          "id",
          "CollectiveId",
          "createdAt",
          "hostCurrency",
          SUM(
            COALESCE("amountInHostCurrency", 0)
            + COALESCE("platformFeeInHostCurrency", 0)
            + COALESCE("hostFeeInHostCurrency", 0)
            + COALESCE("paymentProcessorFeeInHostCurrency", 0)
            + COALESCE("taxAmount" * "hostCurrencyFxRate", 0)
          ) OVER (PARTITION BY "CollectiveId", "hostCurrency" ORDER BY "createdAt" ASC, "id" ASC) as "balance"
          FROM "Transactions"
          WHERE "deletedAt" IS NULL
          AND "CollectiveId" = 69954
          ORDER BY "createdAt" ASC, "id" ASC
```

vs

```
        SELECT
          "id",
          "CollectiveId",
          "createdAt",
          "hostCurrency",
          SUM(
            COALESCE("amountInHostCurrency", 0)
            + COALESCE("platformFeeInHostCurrency", 0)
            + COALESCE("hostFeeInHostCurrency", 0)
            + COALESCE("paymentProcessorFeeInHostCurrency", 0)
            + COALESCE("taxAmount" * "hostCurrencyFxRate", 0)
          ) OVER (PARTITION BY "CollectiveId", "hostCurrency" ORDER BY "createdAt" ASC) as "balance"
          FROM "Transactions"
          WHERE "deletedAt" IS NULL
          AND "CollectiveId" = 69954
          ORDER BY "createdAt" ASC
```

Reply:
We have a sorted index on Transaction's ID column to help with calculating the current balance from the balance checkpoint 
(which uses the max transaction id for a collective). This assumes that we are not back dating transactions from now on, so only id works there.

Here, because we are summing up all transactions and we have some backdated inserts in our history it makes sense to order the sum windows by createdAt.

![image](https://user-images.githubusercontent.com/5448927/210798442-8341037b-c9f8-411a-adfb-2db7f4794e80.png)

Reply:
Thanks for the explanation. I've corrected this part as well. 👍🏽